src/components/ProposalListTable.tsx

```diff
@@ -0,0 +1,116 @@
+import React from 'react';
+import {
+  Table,
+  TableHead,
+  ColumnHead,
+  TableBody,
+  TableRow,
+  RowCell,
+} from './Table';
+
+interface ProposalListTableRowProps {
+  // The `fields` array is currently required, as it determines the column order.
+  // If we can instead recompose the `proposal` children to be in the desired order,
+  // we don't need `fields`.
+  fields: {
+    shortCode: string,
+    label: string,
+  }[],
+  // Note the subtle difference from `proposals` on `ProposalListTableProps`:
+  // this is a single proposal, not an array of them.
+  proposal: {
+    fieldName: string,
+    value: string,
+  }[],
+  proposalId: string,
```

Comment:
> There is no expectation that the proposal’s internal order matches the column (aka field) order, and so there are some potentially expensive iterations in this implementation. We may be able to optimize these if we can prepare the data by putting it in desired order before passing it in.

@reefdog and I discussed this on a call! To recap, I think we should definitely do some data processing to get to the desired format. We should also delegate the responsibility of deciding what whether and in what order to show columns to this display component, while parsing the API response and transforming the data to the expected format should be the responsibility of the page component.

In this case, I think the data might look something like:

```
fieldNames: {
  [shortCode: string]: string;
}
values: {
  [shortCode: string]: string[];
}
```
...that is, we'd pass in two objects: one with the names of canonical fields, and one with arrays of values for those fields, both keyed by the canonical field short codes.

This kind of consolidation is possible here where it isn't on the proposal detail because we don't care about field order in a table with multiple proposals from multiple forms; there does not seem to be an obvious order available.

Reply:
@jasonaowen I think I love that suggestion, but let me make sure I understand it with a pseudo-implementation:

```ts
const fieldNames = {
  [shortCode: string]: string;
}

const proposals = {
  [shortCode: string]: string[];
}[]
```

Right? This is, I think, the correct way to handle the dictionary of field names, being able to access proposal data via a key, and handling multiple values per proposal field.

> …we don't care about field order in a table with multiple proposals from multiple forms; there does not seem to be an obvious order available.

I read this as saying "column order doesn't matter", but it really does! We very much want to arrange the columns in a specific order (which, by the way, will initially involve us hardcoding some shortcodes into the front-end) and eventually by users rearranging them.

But I think that's compatible with your design above, just by adding a `columnOrder` array of shortCodes, the order of which determines the column order. Since we can now access individual fieldNames and proposal attributes using the shortCode key, we're relieved of some performance problems. (Could probably relieve even more by having proposals themselves be arrays of values with no shortCode keys and rearranged as the user changes the column order, but that is premature now and possibly diminishing returns later.)

Reply:
@jasonaowen After re-reading your comment entirely, I realize at the top you directly mention the column ordering, so I definitely misread the bit I quoted. Was that a reference to how field ordering doesn't matter on the detail page?

Reply:
> let me make sure I understand it with a pseudo-implementation: ...

I think you're right; I think I commented on the interface definition for a particular row, but then mixed in the field names which is entirely outside the proposal data! The `proposals` data will need to contain a bit more than that, but I'll comment on the new code rather than get into that here.

> > …we don't care about field order in a table with multiple proposals from multiple forms; there does not seem to be an obvious order available.
> 
> I read this as saying "column order doesn't matter", but it really does! 

Ah sorry, I was referring to the `position` attribute of application form fields. My point is that because we are showing proposals for multiple application forms, I don't think we can order the columns by that `position` attribute, because each application form can and likely will have a different list of fields with different `position`s. Picking any particular application form's field order will necessarily mean showing the fields of proposals to other application forms out-of-order, and that's even before considering that a single application form can have multiple canonical fields with different positions.

I think we're on the same page, though, as you go on to say:

> We very much want to arrange the columns in a specific order (which, by the way, will initially involve us hardcoding some shortCodes into the front-end) and eventually by users rearranging them. But I may be misreading what you wrote because I think you're aware of the column order need.

Yep! Hardcoding is a great first pass, and allowing users to rearrange (and select?) them is exactly what I was imagining and I think we discussed yesterday.

> In any case, though, I think that's still compatible with your design above, just by adding a `columnOrder` array of shortCodes, the order of which determines the column order:
> 
> ```ts
> columnOrder: string[] // Which strings are just `shortCode`s
> ```

Yes, although I would suggest that the column order is the responsibility of the display component, not the API component. (I really need to come up with a better name for these; maybe this is the MVC pattern, with the model being the props, the view being the display component, and the controller being the API component?) That is, the component that parses the API response and transforms the data into what the display component wants should not know or care about the column order - particularly once we add support for re-ordering columns. That does mean the display component will eventually have some state and logic, just none relating to the passed-in data.

> Since we can now access individual fieldNames and proposal attributes using the shortCode key, we're relieved of some performance problems. (Could probably relieve even more by having proposals themselves be arrays of values with no shortCode keys and rearranged as the user changes the column order, but that is premature now and possibly diminishing returns later.)

Yeah, that sounds like an optimization we can consider once we have confirmed we have a performance problem. I'd prefer to start as clearly as possible!